File: src/ui/ai_label_dialog.py
from __future__ import annotations

from typing import TYPE_CHECKING

from PyQt6.QtWidgets import (
    QDialog, QVBoxLayout, QHBoxLayout, QGroupBox, QFormLayout,
    QLineEdit, QPushButton, QDoubleSpinBox, QSpinBox, QRadioButton,
    QCheckBox, QTableWidget, QTableWidgetItem, QComboBox, QLabel,
    QProgressBar, QTextEdit, QDialogButtonBox, QMessageBox, QFileDialog,
)
from PyQt6.QtCore import Qt, QSettings, pyqtSignal
from src.ui.progress_dialog import ProgressDialog, WorkerThread

if TYPE_CHECKING:
    from src.models.project import Project
    from src.core.ai_labeler import AILabeler

_SETTINGS_KEY_MODEL = "ai/model_path"
_SETTINGS_KEY_CONF = "ai/conf_threshold"
_SETTINGS_KEY_CROP = "ai/crop_size"
_SETTINGS_KEY_SCOPE = "ai/scope"
_SETTINGS_KEY_OVERWRITE = "ai/overwrite_labeled"


class ClassMappingTable(QTableWidget):
    def __init__(self, coral_codes: dict[str, str], parent=None) -> None:
        super().__init__(0, 2, parent)
        self._coral_codes = coral_codes
        self.setHorizontalHeaderLabels(["Model Class", "Coral Code"])
        header = self.horizontalHeader()
        if header is not None:
            header.setStretchLastSection(True)
        self.setSelectionMode(QTableWidget.SelectionMode.NoSelection)
        self.setEditTriggers(QTableWidget.EditTrigger.NoEditTriggers)

    def populate(self, suggestions: dict[str, str | None]) -> None:
        self.setRowCount(0)
        options = ["(skip)"] + sorted(self._coral_codes.keys())
        for row, (cls_name, suggested) in enumerate(suggestions.items()):
            self.insertRow(row)
            self.setItem(row, 0, QTableWidgetItem(cls_name))
            combo = QComboBox()
            combo.addItems(options)
            if suggested and suggested in options:
                combo.setCurrentText(suggested)
            self.setCellWidget(row, 1, combo)

    def get_mapping(self) -> dict[str, str | None]:
        result: dict[str, str | None] = {}
        for row in range(self.rowCount()):
            item = self.item(row, 0)
            cls_name = item.text() if item is not None else ""
            widget = self.cellWidget(row, 1)
            value = widget.currentText() if isinstance(widget, QComboBox) else "(skip)"
            result[cls_name] = None if value == "(skip)" else value
        return result


class AILabelDialog(QDialog):
    def __init__(self, project: Project, parent=None) -> None:
        super().__init__(parent)
        self.setWindowTitle("AI Auto-Label")
        self.setMinimumWidth(500)
        self._project = project
        self._labeler: AILabeler | None = None
        settings = QSettings("coralX", "coralX")

        layout = QVBoxLayout(self)

        # --- Model group ---
        model_box = QGroupBox("Model")
        model_form = QFormLayout(model_box)

        path_row = QHBoxLayout()
        self._model_path_edit = QLineEdit()
        self._model_path_edit.setPlaceholderText("Path to YOLOv8 .pt classification model…")
        self._model_path_edit.setText(settings.value(_SETTINGS_KEY_MODEL, ""))
        self._model_path_edit.returnPressed.connect(self._load_model)
        path_row.addWidget(self._model_path_edit)
        browse_btn = QPushButton("Browse…")
        browse_btn.clicked.connect(self._browse_model)
        path_row.addWidget(browse_btn)
        model_form.addRow("Model file (.pt):", path_row)

        self._conf_spin = QDoubleSpinBox()
        self._conf_spin.setRange(0.0, 1.0)
        self._conf_spin.setSingleStep(0.05)
        self._conf_spin.setDecimals(2)
        self._conf_spin.setValue(float(settings.value(_SETTINGS_KEY_CONF, 0.5)))
        model_form.addRow("Confidence threshold:", self._conf_spin)

        self._crop_spin = QSpinBox()
        self._crop_spin.setRange(32, 512)
        self._crop_spin.setSingleStep(32)
        self._crop_spin.setValue(int(settings.value(_SETTINGS_KEY_CROP, 64)))
        self._crop_spin.setSuffix(" px")
        model_form.addRow("Crop size:", self._crop_spin)

        layout.addWidget(model_box)

        # --- Scope group ---
        scope_box = QGroupBox("Scope")
        scope_layout = QVBoxLayout(scope_box)
        self._scope_image = QRadioButton("This image only")
        self._scope_station = QRadioButton("This station")
        self._scope_project = QRadioButton("Entire project")
        saved_scope = settings.value(_SETTINGS_KEY_SCOPE, "image")
        if saved_scope == "station":
            self._scope_station.setChecked(True)
        elif saved_scope == "project":
            self._scope_project.setChecked(True)
        else:
            self._scope_image.setChecked(True)
        scope_layout.addWidget(self._scope_image)
        scope_layout.addWidget(self._scope_station)
        scope_layout.addWidget(self._scope_project)
        layout.addWidget(scope_box)

        # --- Overwrite checkbox ---
        self._overwrite_cb = QCheckBox("Label only unlabeled points (uncheck to overwrite all)")
        saved_overwrite = settings.value(_SETTINGS_KEY_OVERWRITE, "true")
        self._overwrite_cb.setChecked(saved_overwrite != "false")
        layout.addWidget(self._overwrite_cb)

        # --- Class mapping group ---
        mapping_box = QGroupBox("Class Mapping")
        mapping_layout = QVBoxLayout(mapping_box)
        self._hint_label = QLabel("Select a .pt model file above to populate this table.")
        self._hint_label.setStyleSheet("color: gray; font-style: italic;")
        mapping_layout.addWidget(self._hint_label)
        self._mapping_table = ClassMappingTable(project.coral_codes)
        self._mapping_table.setMinimumHeight(120)
        self._mapping_table.hide()
        mapping_layout.addWidget(self._mapping_table)
        reload_btn = QPushButton("Load Model")
        reload_btn.clicked.connect(self._load_model)
        mapping_layout.addWidget(reload_btn)
        layout.addWidget(mapping_box)

        # --- Buttons ---
        self._btn_box = QDialogButtonBox()
        run_btn = self._btn_box.addButton("Run", QDialogButtonBox.ButtonRole.AcceptRole)
        assert run_btn is not None
        self._run_btn: QPushButton = run_btn
        self._btn_box.addButton(QDialogButtonBox.StandardButton.Cancel)
        self._run_btn.setEnabled(False)
        self._btn_box.accepted.connect(self.accept)
        self._btn_box.rejected.connect(self.reject)
        layout.addWidget(self._btn_box)

    def _browse_model(self) -> None:
        path, _ = QFileDialog.getOpenFileName(
            self, "Select YOLOv8 Model", "", "YOLO model (*.pt)"
        )
        if path:
            self._model_path_edit.setText(path)
            self._load_model()

    def _load_model(self) -> None:
        from src.core.ai_labeler import AILabeler  # pylint: disable=import-outside-toplevel

        path = self._model_path_edit.text().strip()
        if not path:
            QMessageBox.warning(self, "No model selected", "Please select a .pt model file first.")
            return

        _project = self._project

        def _run(cb):
            cb(0, 0, "Loading AI model…")
            labeler = AILabeler(path)
            cb(0, 0, "Mapping classes to coral codes…")
            suggestions = AILabeler.suggest_mapping(
                labeler.class_names(), _project.coral_codes
            )
            return labeler, suggestions

        def _on_done(result):
            dlg.accept()
            labeler, suggestions = result
            self._mapping_table.populate(suggestions)
            self._hint_label.hide()
            self._mapping_table.show()
            self._labeler = labeler
            self._run_btn.setEnabled(True)
            task_label = "detection" if labeler.task == "detect" else "classification"
            QMessageBox.information(
                self, "Model Loaded",
                f"Model loaded successfully.\n"
                f"Type: {task_label}\n"
                f"Classes ({len(labeler.class_names())}): {', '.join(labeler.class_names())}",
            )

        def _on_error(msg):
            dlg.accept()
            QMessageBox.critical(self, "Failed to Load Model", msg)

        dlg = ProgressDialog("Loading AI Model…", cancellable=False, parent=self)
        dlg.set_indeterminate("Loading model, please wait…")
        worker = WorkerThread(_run, parent=self)
        worker.progress.connect(dlg.update)
        worker.succeeded.connect(_on_done)
        worker.failed.connect(_on_error)
        worker.start()
        dlg.exec()

    def accept(self) -> None:
        settings = QSettings("coralX", "coralX")
        settings.setValue(_SETTINGS_KEY_MODEL, self._model_path_edit.text())
        settings.setValue(_SETTINGS_KEY_CONF, self._conf_spin.value())
        settings.setValue(_SETTINGS_KEY_CROP, self._crop_spin.value())
        settings.setValue(_SETTINGS_KEY_SCOPE, self.scope())
        overwrite_val = "true" if self._overwrite_cb.isChecked() else "false"
        settings.setValue(_SETTINGS_KEY_OVERWRITE, overwrite_val)
        super().accept()

    # --- Public properties ---

    def model_path(self) -> str:
        return self._model_path_edit.text().strip()

    def conf_threshold(self) -> float:
        return self._conf_spin.value()

    def crop_size(self) -> int:
        return self._crop_spin.value()

    def scope(self) -> str:
        if self._scope_station.isChecked():
            return "station"
        if self._scope_project.isChecked():
            return "project"
        return "image"

    def overwrite_labeled(self) -> bool:
        return not self._overwrite_cb.isChecked()

    def class_mapping(self) -> dict[str, str | None]:
        return self._mapping_table.get_mapping()

    @property
    def labeler(self) -> AILabeler | None:
        return self._labeler


class AIProgressDialog(QDialog):
    cancel_requested = pyqtSignal()

    def __init__(self, total_points: int, parent=None) -> None:
        super().__init__(parent)
        self.setWindowTitle("AI Auto-Label — Running")
        self.setMinimumWidth(480)
        self.setMinimumHeight(300)
        self.setWindowModality(Qt.WindowModality.ApplicationModal)

        layout = QVBoxLayout(self)

        layout.addWidget(QLabel("Running AI inference…"))

        self._status_label = QLabel("Starting…")
        layout.addWidget(self._status_label)

        self._progress_bar = QProgressBar()
        self._progress_bar.setRange(0, total_points)
        self._progress_bar.setValue(0)
        layout.addWidget(self._progress_bar)

        self._log = QTextEdit()
        self._log.setReadOnly(True)
        self._log.setMinimumHeight(120)
        font = self._log.font()
        font.setFamily("Monospace")
        font.setPointSize(9)
        self._log.setFont(font)
        layout.addWidget(self._log)

        self._cancel_btn = QPushButton("Cancel")
        self._cancel_btn.clicked.connect(self._on_cancel)
        layout.addWidget(self._cancel_btn, alignment=Qt.AlignmentFlag.AlignRight)

    def _on_cancel(self) -> None:
        if self._cancel_btn.text() == "Cancel":
            self.cancel_requested.emit()
            self._cancel_btn.setEnabled(False)
            self._status_label.setText("Cancelling…")
        else:
            self.accept()

    def on_progress(self, done: int, _total: int, status: str) -> None:
        self._progress_bar.setValue(done)
        self._status_label.setText(status)
        self._log.append(f"  {status}")
        # Auto-scroll to bottom
        scrollbar = self._log.verticalScrollBar()
        if scrollbar is not None:
            scrollbar.setValue(scrollbar.maximum())

    def on_error(self, msg: str) -> None:
        self._log.append(f"<span style='color:red'>ERROR: {msg}</span>")
        self._cancel_btn.setEnabled(True)

    def on_finished(self) -> None:
        self._cancel_btn.setText("Close")
        self._cancel_btn.setEnabled(True)
        self._status_label.setText("Done.")

File: rust/src/ui/ai_label_dialog.rs
//! AI auto-label configuration dialog and run-progress dialog.

use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::mpsc::Receiver;
use std::sync::Arc;

use egui::{Color32, Context};

use crate::core::ai_labeler::{self, AiLabeler, LabelResult, WorkerEvent};
use crate::models::ImageAnnotation;
use crate::ui::progress_dialog::{spawn_worker, WorkerEvent as SimpleEvent};
use crate::ui::settings::AiSettings;

#[derive(PartialEq, Clone, Copy)]
pub enum Scope {
    Image,
    Station,
    Project,
}

impl Scope {
    fn as_str(self) -> &'static str {
        match self {
            Scope::Image => "image",
            Scope::Station => "station",
            Scope::Project => "project",
        }
    }
    fn from_str(s: &str) -> Self {
        match s {
            "station" => Scope::Station,
            "project" => Scope::Project,
            _ => Scope::Image,
        }
    }
}

enum ModelLoad {
    None,
    Loading(Receiver<SimpleEvent<(AiLabeler, HashMap<String, Option<String>>)>>),
    Loaded,
}

pub struct AiLabelDialog {
    model_path: String,
    conf_threshold: f64,
    crop_size: i64,
    scope: Scope,
    /// Checkbox state: "Label only unlabeled points" — the inverse of the
    /// `overwrite_labeled` flag the labeling job actually wants.
    only_unlabeled: bool,
    coral_codes: HashMap<String, String>,
    class_mapping: Vec<(String, Option<String>)>,
    labeler: Option<AiLabeler>,
    load: ModelLoad,
    load_error: Option<String>,
    info_message: Option<String>,
}

pub struct RunConfig {
    pub labeler: AiLabeler,
    pub class_mapping: HashMap<String, Option<String>>,
    pub conf_threshold: f32,
    pub crop_size: u32,
    pub overwrite_labeled: bool,
    pub scope: &'static str,
}

impl AiLabelDialog {
    pub fn new(coral_codes: HashMap<String, String>) -> Self {
        let s = AiSettings::load();
        AiLabelDialog {
            model_path: s.model_path,
            conf_threshold: s.conf_threshold,
            crop_size: s.crop_size,
            scope: Scope::from_str(&s.scope),
            only_unlabeled: s.only_unlabeled,
            coral_codes,
            class_mapping: Vec::new(),
            labeler: None,
            load: ModelLoad::None,
            load_error: None,
            info_message: None,
        }
    }

    fn save_settings(&self) {
        AiSettings {
            model_path: self.model_path.clone(),
            conf_threshold: self.conf_threshold,
            crop_size: self.crop_size,
            scope: self.scope.as_str().to_string(),
            only_unlabeled: self.only_unlabeled,
        }
        .save();
    }

    fn start_load(&mut self) {
        let path = self.model_path.trim().to_string();
        if path.is_empty() {
            self.load_error = Some("Please select a .pt/.onnx model file first.".to_string());
            return;
        }
        let coral_codes = self.coral_codes.clone();
        let rx = spawn_worker(move |_cb| {
            let labeler = AiLabeler::load(&path)?;
            let suggestions = AiLabeler::suggest_mapping(labeler.class_names(), &coral_codes);
            Ok((labeler, suggestions))
        });
        self.load = ModelLoad::Loading(rx);
    }

    /// Draws the dialog; returns `Some(config)` once the user clicks Run with
    /// a model loaded.
    pub fn show(&mut self, ctx: &Context) -> Option<RunConfig> {
        // Poll the background model-load job, if any.
        if let ModelLoad::Loading(rx) = &self.load {
            match rx.try_recv() {
                Ok(SimpleEvent::Succeeded((labeler, suggestions))) => {
                    let mut options: Vec<String> = self.coral_codes.keys().cloned().collect();
                    options.sort();
                    self.class_mapping = suggestions
                        .into_iter()
                        .map(|(cls, suggested)| (cls, suggested.filter(|s| options.contains(s))))
                        .collect();
                    self.info_message = Some(format!(
                        "Model loaded successfully.\nType: {}\nClasses ({}): {}",
                        if labeler.task() == "detect" { "detection" } else { "classification" },
                        labeler.class_names().len(),
                        labeler.class_names().join(", ")
                    ));
                    self.labeler = Some(labeler);
                    self.load = ModelLoad::Loaded;
                }
                Ok(SimpleEvent::Failed(msg)) => {
                    self.load_error = Some(msg);
                    self.load = ModelLoad::None;
                }
                Ok(SimpleEvent::Progress { .. }) | Err(std::sync::mpsc::TryRecvError::Empty) => {}
                Err(std::sync::mpsc::TryRecvError::Disconnected) => self.load = ModelLoad::None,
            }
        }

        let mut run_config = None;

        egui::Window::new("AI Auto-Label").collapsible(false).min_width(500.0).show(ctx, |ui| {
            ui.group(|ui| {
                ui.label("Model");
                ui.horizontal(|ui| {
                    ui.label("Model file (.pt/.onnx):");
                    ui.text_edit_singleline(&mut self.model_path);
                    if ui.button("Browse...").clicked() {
                        if let Some(path) = rfd::FileDialog::new().add_filter("YOLO model", &["pt", "onnx"]).pick_file() {
                            self.model_path = path.to_string_lossy().into_owned();
                            self.start_load();
                        }
                    }
                });
                ui.horizontal(|ui| {
                    ui.label("Confidence threshold:");
                    ui.add(egui::Slider::new(&mut self.conf_threshold, 0.0..=1.0).step_by(0.05));
                });
                ui.horizontal(|ui| {
                    ui.label("Crop size:");
                    ui.add(egui::Slider::new(&mut self.crop_size, 32..=512).step_by(32.0).suffix(" px"));
                });
            });

            ui.group(|ui| {
                ui.label("Scope");
                ui.radio_value(&mut self.scope, Scope::Image, "This image only");
                ui.radio_value(&mut self.scope, Scope::Station, "This station");
                ui.radio_value(&mut self.scope, Scope::Project, "Entire project");
            });

            ui.checkbox(&mut self.only_unlabeled, "Label only unlabeled points (uncheck to overwrite all)");

            ui.group(|ui| {
                ui.label("Class Mapping");
                if matches!(self.load, ModelLoad::Loading(_)) {
                    ui.horizontal(|ui| {
                        ui.spinner();
                        ui.label("Loading model...");
                    });
                } else if self.class_mapping.is_empty() {
                    ui.colored_label(Color32::GRAY, "Select a model file above to populate this table.");
                } else {
                    let mut options: Vec<String> = self.coral_codes.keys().cloned().collect();
                    options.sort();
                    egui::Grid::new("class_mapping_grid").striped(true).show(ui, |ui| {
                        ui.label("Model Class");
                        ui.label("Coral Code");
                        ui.end_row();
                        for (cls, selected) in &mut self.class_mapping {
                            ui.label(cls.as_str());
                            let text = selected.clone().unwrap_or_else(|| "(skip)".to_string());
                            egui::ComboBox::from_id_salt(format!("map_{cls}")).selected_text(text).show_ui(ui, |ui| {
                                if ui.selectable_label(selected.is_none(), "(skip)").clicked() {
                                    *selected = None;
                                }
                                for code in &options {
                                    if ui.selectable_label(selected.as_deref() == Some(code.as_str()), code).clicked() {
                                        *selected = Some(code.clone());
                                    }
                                }
                            });
                            ui.end_row();
                        }
                    });
                }
                if ui.button("Load Model").clicked() {
                    self.start_load();
                }
            });

            if let Some(err) = &self.load_error {
                ui.colored_label(Color32::from_rgb(0xd6, 0x27, 0x28), err);
            }
            if let Some(info) = self.info_message.clone() {
                ui.colored_label(Color32::from_rgb(0x6a, 0xcc, 0x65), &info);
            }

            ui.separator();
            ui.with_layout(egui::Layout::right_to_left(egui::Align::Center), |ui| {
                let can_run = self.labeler.is_some();
                if ui.add_enabled(can_run, egui::Button::new("Run")).clicked() {
                    self.save_settings();
                    if let Some(labeler) = self.labeler.take() {
                        let class_mapping: HashMap<String, Option<String>> = self.class_mapping.iter().cloned().collect();
                        run_config = Some(RunConfig {
                            labeler,
                            class_mapping,
                            conf_threshold: self.conf_threshold as f32,
                            crop_size: self.crop_size as u32,
                            overwrite_labeled: !self.only_unlabeled,
                            scope: self.scope.as_str(),
                        });
                    }
                }
                if ui.button("Cancel").clicked() {
                    self.save_settings();
                }
            });
        });

        run_config
    }
}

/// Live progress + scrolling log while an AI labeling job runs in the background.
pub struct AiProgressDialog {
    rx: Receiver<WorkerEvent>,
    cancel: Arc<AtomicBool>,
    done: usize,
    total: usize,
    status: String,
    log: Vec<String>,
    finished: bool,
    result: Vec<LabelResult>,
}

impl AiProgressDialog {
    pub fn spawn(
        labeler: AiLabeler,
        annotations: Vec<(String, ImageAnnotation)>,
        class_mapping: HashMap<String, Option<String>>,
        conf_threshold: f32,
        crop_size: u32,
        overwrite_labeled: bool,
    ) -> Self {
        let total: usize = annotations
            .iter()
            .map(|(_, a)| a.points.iter().filter(|p| overwrite_labeled || p.label.is_none()).count())
            .sum();
        let (rx, cancel) = ai_labeler::spawn_label_worker(labeler, annotations, class_mapping, conf_threshold, crop_size, overwrite_labeled);
        AiProgressDialog { rx, cancel, done: 0, total, status: "Starting...".to_string(), log: Vec::new(), finished: false, result: Vec::new() }
    }

    pub fn is_finished(&self) -> bool {
        self.finished
    }

    pub fn take_result(self) -> Vec<LabelResult> {
        self.result
    }

    pub fn show(&mut self, ctx: &Context) {
        loop {
            match self.rx.try_recv() {
                Ok(WorkerEvent::Progress { done, total, message }) => {
                    self.done = done;
                    self.total = total;
                    self.status = message.clone();
                    self.log.push(message);
                }
                Ok(WorkerEvent::Error(msg)) => {
                    self.log.push(format!("ERROR: {msg}"));
                }
                Ok(WorkerEvent::Finished(results)) => {
                    self.result = results;
                    self.finished = true;
                    self.status = "Done.".to_string();
                    break;
                }
                Err(_) => break,
            }
        }

        egui::Window::new("AI Auto-Label — Running").collapsible(false).min_width(480.0).min_height(300.0).show(ctx, |ui| {
            ui.label("Running AI inference...");
            ui.label(&self.status);
            ui.add(egui::ProgressBar::new(self.done as f32 / self.total.max(1) as f32).show_percentage());

            egui::ScrollArea::vertical().max_height(150.0).stick_to_bottom(true).show(ui, |ui| {
                for line in &self.log {
                    ui.monospace(line);
                }
            });

            ui.with_layout(egui::Layout::right_to_left(egui::Align::Center), |ui| {
                if self.finished {
                    if ui.button("Close").clicked() {
                        // caller checks is_finished() + closes
                    }
                } else {
                    let cancelling = self.cancel.load(Ordering::Relaxed);
                    if ui.add_enabled(!cancelling, egui::Button::new(if cancelling { "Cancelling..." } else { "Cancel" })).clicked() {
                        self.cancel.store(true, Ordering::Relaxed);
                        self.status = "Cancelling...".to_string();
                    }
                }
            });
        });
        ctx.request_repaint();
    }
}
